### PR TITLE
797 - Generate and add `metadata.tsv` to a zip file

### PR DESCRIPTION
**Pull request by nozomione**
## Issue Number
#797

## Purpose/Implementation Notes
This is the 3rd item in the stacked PR list for the linked issue.

The following changes are included:

- Added the logic for the `metadata.tsv` file generation to `ComputedFile::get_portal_metadata_file` 
- Add the static method `ComputedFile::get_local_portal_metadata_path:` and remove  `common.OUTPUT_PORTAL_METADATA_ZIP_FILE_PATH` 
- Update the `create-portal-metadata` command and its test

## Types of changes

- New feature (non-breaking change which adds functionality)

## Functional tests
N/A

## Checklist
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Screenshots
N/A

### CODE REVIEW

**Comment by nozomione** [comment]
I've removed `TODO` comments and adjusted the codeabse based on 1:1, as well as updated the "Purpose/Implementation Notes" section of this PR.

Changes includes:
- Removed`TODO` and adjusted the codebase using the new readme folder structure (with #831)
- Removed the individual test method (metadata tsv) and directly asserted in `test_create_portal_metadata`
- Used the `csv` module for rows check

This is ready for another look, thank you David!

(**NOTE:** minor adjustment were made for better readability at b2d26cbd67edc9a7a6d93200692e49b8e837a1b2.)

File: api/scpca_portal/models/computed_file.py
Comment (davidsmejia) [inline api/scpca_portal/models/computed_file.py:105]:
```suggestion
        # If the query returns empty, then an error occurred, and we should abort early
```

File: api/scpca_portal/models/computed_file.py
Comment (davidsmejia) [inline api/scpca_portal/models/computed_file.py:94]:
```suggestion
    @staticmethod
    def get_local_file_path(download_config: Dict) -> Path:
        """ Takes a download_config dictionary and returns the filepath where the zipfile will be saved locally before upload. """
        if download_config is common.GENERATED_PORTAL_METADATA_DOWNLOAD_CONFIG:
            return common.OUTPUT_DATA_PATH / common.PORTAL_METADATA_COMPUTED_FILE_NAME
```

We should consolidate these and standardize how we get the value. For now let's keep how project and sample computed file names/ paths are evaluated on their models. We can move them under here later.

File: api/scpca_portal/models/computed_file.py
Comment (davidsmejia) [inline api/scpca_portal/models/computed_file.py:114]:
```suggestion
        zip_file_path = cls.get_local_file_path(download_config)
        with ZipFile(zip_file_path, "w") as zip_file:
```

File: api/scpca_portal/models/computed_file.py
Comment (davidsmejia) [inline api/scpca_portal/models/computed_file.py:134]:
```suggestion
            size_in_bytes=zip_file_path.stat().st_size,
```

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:18]:
```suggestion
```

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:57]:
```suggestion
        zip_file_path = ComputedFile.get_local_file_path(common.GENERATED_PORTAL_METADATA_DOWNLOAD_CONFIG)
        with ZipFile(zip_file_path) as zip:
```

Comment (davidsmejia) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py:57]: This would be the preferred way to handle this for now. In the future we should persist the download_config to the computed file instance so we can query the computed file and then grab the local path as an attribute.

**Comment by nozomione** [comment]
I've applied your feedback and this PR is now ready for another look, thank you David!

File: api/scpca_portal/models/computed_file.py
Comment (davidsmejia) [inline api/scpca_portal/models/computed_file.py:109]:
```suggestion
        if not libraries.exists():
            logger.error("There are no libraries on the portal!")
```

File: api/scpca_portal/models/computed_file.py
Comment (davidsmejia) [inline api/scpca_portal/models/computed_file.py:136]:
```suggestion
```

This makes sense to keep in common since it is a special case output file that does not belong to any specific project or sample.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:68]:
```suggestion
```

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:71]: We can remove both of these checks since we are already asserting that they are in the zip.namelist immediately after this. You can assert that it is 2 files if you want to ensure there are no extraneous files in the output zip.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:73]:
```suggestion
            self.assertIn(readme_file.OUTPUT_NAME, files)
            self.assertIn(readme_file.METADATA_ONLY_FILE_NAME, files)
```

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:86]: You should be able to open this with a combination of TextIOWrapper and csv.DictReader. Then you just need to ensure that the keys from the first object is a subset of `METADATA_COLUMN_SORT_ORDER`.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:89]: The other test would be to pull out all of the library ids from the parsed metadata file and ensure that the DB does not contain any extra.

```python
present_libraries = set([metadata_dict['scpca_library_id'] for metadata_dict in metadata_dicts])
all_libraries = set(Library.objects.all().values_list('scpca_library_id', flat=True))
self.assertEquals(preset_libraries, all_libraries)
```

Comment (davidsmejia) [inline api/scpca_portal/test/management/commands/test_create_portal_metadata.py:89]:
```suggestion
            present_keys = set(metadata_dicts[0].keys())
            allowed_keys = set(common.METADATA_COLUMN_SORT_ORDER) - set(['*'])
            self.assertEqual(present_keys, allowed_keys)
```

Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: See at 651d753c49c931eb71237155e64ad99226d6e9c8.

File: api/scpca_portal/models/computed_file.py
Comment (nozomione) [reply api/scpca_portal/models/computed_file.py:109]: See at 651d753c49c931eb71237155e64ad99226d6e9c8.

File: api/scpca_portal/models/computed_file.py
Comment (nozomione) [reply api/scpca_portal/models/computed_file.py]: See at 651d753c49c931eb71237155e64ad99226d6e9c8.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: See at 651d753c49c931eb71237155e64ad99226d6e9c8.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: See at 651d753c49c931eb71237155e64ad99226d6e9c8.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: See at 651d753c49c931eb71237155e64ad99226d6e9c8.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: See at 651d753c49c931eb71237155e64ad99226d6e9c8.

File: api/scpca_portal/test/management/commands/test_create_portal_metadata.py
Comment (nozomione) [reply api/scpca_portal/test/management/commands/test_create_portal_metadata.py]: See at 651d753c49c931eb71237155e64ad99226d6e9c8.

**Comment by nozomione** [comment]
The test is now more comprehensive, including additional assertions (e.g., checking the count of library Ids). Thank you for your great insights. This PR is now ready for another review, thank you David!

**Review by davidsmejia — APPROVED** [review]
LGTM, but let's merge after 808 is approved.

**Comment by nozomione** [comment]
PR #808 has been approved and merged into the feature branch `feature/portal-metadata-command`, so we are now merging this PR into it as well.

**Pull request merged**
